### PR TITLE
RFC: Have each process write its own unique .cov file for coverage statistics

### PR DESCRIPTION
**Pull request by kshyatt**
Right now, files from multiple processes (e.g. for `parallel` and `multi` tests) are overwriting each other and throwing our statistics off. This PR changes the behaviour so that when `--code-coverage={user/all}` is passed, each process will write its own file of the format `{filename}.{pid}.cov`. I've already tested this with `Coverage.jl` and `CoverageBase.jl` and have submitted PRs to handle the change to those two packages.

### CODE REVIEW

**Comment by kshyatt** [comment]
Oh no test failures! My Linux and OSX environments are running with new enough compilers that C++11 is the default so that's why a bunch of these failed (I was lazy and used `std::to_string`). Somehow I also passed a local version of the `pkg` tests that failed on OSX here. I'll see if I can hunt that down.

**Comment by KristofferC** [comment]
Should this be activated by default or should there be some setting to activate it? I am thinking that most users run their testing in a single process and it might add some unnecessary clutter to print the pids all the time.

Also does the default `.gitignore` that is generated by `Pkg.generate` need to be tweaked?

**Comment by tkelman** [comment]
Judging by the number of bugs that happen only when you run the tests in a single process that have gone unnoticed on master every so often, I think more people run the base tests in parallel than otherwise.

When you run `Pkg.test("Foo", coverage=true)` it will also generate `.cov` files in the package directory, so makes sense to me to add them to gitignore.

**Comment by KristofferC** [comment]
I was thinking more about when users run local coverage testing on their packages and getting a bunch of new .cov files every time the test is run. Maybe it is not a big deal.

**Comment by timholy** [comment]
You kind of already need a nice way to do this, or it makes grep results messier. Locally I've defined `cleancovmem`:
```sh
#! /bin/bash
fls=$(find . -name "*.jl.cov" -o -name "*.jl.mem")
rm $fls
```
After this PR one will need an extra `*.` in those filenames.

**Comment by ScottPJones** [comment]
Is there a way to have the .cov* and .mem* files written to another directory?
I know that there is already a move to use CMake, and not dump the .o files in the source directories,
it would be very nice to have out-of-source coverage and profiling also.

**Comment by kshyatt** [comment]
So AFAIK we are now using by default on every platform a version of LLVM that supports `-std=c++11`, but `Make.inc` only compiles with that flag if you are using `llvm-svn`. That was set in May 2014. Most of these CI runs failed because I didn't have that flag - should we think about modifying the `Make.inc` file, or should I just give up and `sprintf`?

**Comment by tkelman** [comment]
Linux Travis is using gcc 4.6 (ubuntu 12.04) which I don't think supports `-std=c++11`

**Comment by kshyatt** [comment]
It might support `-std=c++0x`. But whatever, I'll `ostringstream` instead :).

File: test/choosetests.jl [hunk 1]
```diff
@@ -30,7 +30,7 @@ function choosetests(choices = [])
         "replutil", "sets", "test", "goto", "llvmcall", "grisu",
         "nullable", "meta", "profile", "libgit2", "docs", "markdown",
         "base64", "parser", "serialize", "functors", "char", "misc",
-        "enums", "cmdlineargs", "i18n", "workspace"
+        "enums", "cmdlineargs", "i18n", "workspace", "pkg"
```
Comment (tkelman) [inline test/choosetests.jl]: can't do this, distribution buildbots don't have internet access

Comment (kshyatt) [inline test/choosetests.jl]: Curses! I'll remove that.

**Comment by kshyatt** [comment]
Ok, the AV tests got owned by my `test/pkg` inclusion. All other tests on AV passed. We'll see how the Linux and OSX tests do and then I'll rebase.

**Comment by yuyichao** [comment]
I believe the travis tests include pkg.

**Comment by tkelman** [comment]
wat? `git: 'pull' is not a git command. See 'git --help'.` Messed-up shell quoting maybe? https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.6543/job/fxsxl0nd3imh20lt#L1111

I think the pkg test should theoretically work on appveyor, but we don't run it at the moment out of run time / git binary fiddling concerns.

**Comment by kshyatt** [comment]
> I believe the travis tests include pkg.
I should have checked the `travis.yml` file. At least I'm not dying to C++11 not being supported anymore.

**Comment by yuyichao** [comment]
http://stackoverflow.com/questions/1465398/git-pull-broken

Path issue?

**Comment by tkelman** [comment]
> Path issue?

Yeah, likely. I'm copying over pieces of the command-line git from the nightly binary but not all of it, so I'm not surprised it's slightly broken. Unrelated to this PR, anyway.

**Comment by kshyatt** [comment]
OK I am going to rebase now and remove the offending error in `test/choosetests.jl`.

**Comment by kshyatt** [comment]
Does anyone know what's up with that one failing OSX run? If it's not my fault, is this ok to merge?

**Comment by tkelman** [comment]
I've seen that segfault elsewhere, we should open a separate issue on it, unrelated to this PR. Merge away.

**Pull request merged**